### PR TITLE
[bugfix] Check LightningOptimizer doesn't delete optimizer hooks

### LINKED ISSUE #5131: Updating 1.0.3 -> 1.0.4 breaks pytorch module hooks
#### What is your question?

I implemented a custom optimizer that stores the input and the incoming gradients for each layer and recomputes the gradients in a per-sample fashion (wich I do because I want to use K-FAC for my optimization).

I'm using the pytorch hooks on modules (i.e. module123.register_forward_pre_hook(_save_input_function) which will store the inputs into the state dictionary of the optimizer.
I use the same mod123.register_backward_hook(_save_gradient_function) for the backward gradients) and when calling optimizer.step().

As I updated from 1.0.1 to 1.1.0 it broke the hooks that are required in the KFAC optimizer.
I checked different versions and from 1.0.3 to 1.0.4 something changed such that the Trainer ignores/steps over the hooks in the forward and backward pass.

I checked the release notes but couldn't identify anything essential that would warrant these breaking changes.

Additionally they were issue popping up with manually modifying gradients in other sampling-based types of optimizers.

These issue do not exist for the standard optimizers straight from PyTorch optim class.

#### Code
It would be a huge codebase as the implementation of KFAC is a bit sophisticated.
But if nothing helps I could write an minimal working example from scratch.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a test to make sure hooks aren't Optimizer being deleted when LightningOptimizer is deleted

Fixes #5131

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_lightning_optimizer.py [hunk 1]
```diff
@@ -303,3 +306,100 @@ def configure_optimizers(self):
...
+        closure()
+        for group in self.param_groups:
+            _ = self.state[group['mod']]['x']
+            _ = self.state[group['mod']]['grad']
+        return True
+
+
+def test_lightning_optimizer_dont_delete_wrapped_optimizer(tmpdir):
+    """
+    """
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:375]:
```suggestion
```

File: tests/core/test_lightning_optimizer.py [hunk 2]
```diff
@@ -303,3 +306,100 @@ def configure_optimizers(self):
...
+                self._bwd_handles.append(handle)  # collect backward-save-grad hook in list
+
+                # save the parameters
+                params = [mod.weight]
+                if mod.bias is not None:
+                    params.append(mod.bias)
+
+                # save a param_group for each module
+                d = {'params': params, 'mod': mod, 'layer_type': mod_class}
+                self.params.append(d)
```
Comment (Borda) [inline tests/core/test_lightning_optimizer.py:347]:
```suggestion
            if mod_class not in ['Linear']:  # silently skips other layers
                continue

            # save the inputs and gradients for the kfac matrix computation
            handle = mod.register_forward_pre_hook(self._save_input)  # save the inputs
            self._fwd_handles.append(handle)  # collect forward-save-input hooks in list
            handle = mod.register_backward_hook(self._save_grad_output)  # save the gradients
            self._bwd_handles.append(handle)  # collect backward-save-grad hook in list

            # save the parameters
            params = [mod.weight]
            if mod.bias is not None:
                params.append(mod.bias)

            # save a param_group for each module
            d = {'params': params, 'mod': mod, 'layer_type': mod_class}
            self.params.append(d)
```

**Review by Borda — APPROVED** [review]

File: tests/core/test_lightning_optimizer.py [hunk 3]
```diff
@@ -303,3 +306,100 @@ def configure_optimizers(self):
...
+        def on_train_batch_start(self, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+            self.count_on_train_batch_start += 1
+            optimizer = self.optimizers(use_pl_optimizer=False)
+            assert len(optimizer._fwd_handles) == 1
+
+        def on_train_batch_end(self, outputs: Any, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+            self.count_on_train_batch_end += 1
+            # delete the lightning_optimizers
+            self.trainer._lightning_optimizers = None
+            gc.collect()
```
Comment (carmocca) [inline tests/core/test_lightning_optimizer.py:396]: does the test not pass without this?

Comment (tchaton) [reply tests/core/test_lightning_optimizer.py:395]: Yes, it does, but just in case. It is related the issue message about del LightningOptimizer + `gc.collect()` delete hooks.

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]
